dace/sdfg/propagation.py

```diff
@@ -546,16 +553,346 @@ def propagate(self, array, expressions, node_range):
+                            'proposed_dynamic': proposed_dynamic,
+                        })
+
+        if len(remaining_branches) > 0:
+            # Traverse down next remaining branch.
+            branch = remaining_branches.pop()
+            state = branch['state']
+            proposed_executions = branch['proposed_executions']
+            proposed_dynamic = branch['proposed_dynamic']
+        else:
+            # No more remaining branch, done.
+            state = None
+            break
+
+
+def propagate_memlets_sdfg(sdfg, border_memlets=None):
     """ Propagates memlets throughout an entire given SDFG. 
+        :param border_memlets: A map between connectors to and from this SDFG
+                               (if it's a nested SDFG), to their connected
+                               'border' memlets.
         :note: This is an in-place operation on the SDFG.
     """
     for state in sdfg.nodes():
         propagate_memlets_state(sdfg, state)
 
+    propagate_states(sdfg)
+
+    if border_memlets is not None:
+        # For each state, go through all access nodes corresponding to any in-
+        # or out-connectors to and from this SDFG. Given those access nodes,
+        # collect the corresponding memlets and use them to calculate the
+        # memlet volume and subset corresponding to the outside memlet attached
+        # to that connector. This is passed out via `border_memlets` and
+        # propagated along from there.
+        for state in sdfg.nodes():
+            for node in state.nodes():
+                if isinstance(node, AccessNode):
+                    for direction in border_memlets:
+                        if (node.label in border_memlets[direction]):
+                            memlet = border_memlets[direction][node.label]
+
+                            # Collect the edges to/from this access node,
+                            # depending on the direction the connector leads in.
+                            edges = []
+                            if direction == 'in':
+                                edges = state.out_edges(node)
+                            elif direction == 'out':
+                                edges = state.in_edges(node)
+
+                            # Collect all memlets belonging to this access node,
+                            # and accumulate the total volume between them.
+                            memlets = []
+                            for edge in edges:
+                                inside_memlet = edge.data
+                                memlets.append(inside_memlet)
+
+                                if memlet is None:
+                                    # Use the first encountered memlet as a
+                                    # 'border' memlet and accumulate the sum
+                                    # on it.
+                                    memlet = copy.deepcopy(inside_memlet)
```

Comment:
at this point, just create a new memlet with only `data` set :) why even copy?

Reply:
Addressed in 9afab7c